core/matrix/dense.cpp

```diff
@@ -461,6 +470,35 @@ void Dense<ValueType>::compute_norm1_impl(LinOp* result) const
 }
 
 
+template <typename ValueType>
+void Dense<ValueType>::compute_squared_norm2(LinOp* result,
+                                             array<char>& tmp) const
+{
+    GKO_ASSERT_EQUAL_DIMENSIONS(result, dim<2>(1, this->get_size()[1]));
+    auto exec = this->get_executor();
+    if (tmp.get_executor() != exec) {
+        tmp.clear();
+        tmp.set_executor(exec);
+    }
+    auto local_result = make_temporary_clone(exec, result);
+    auto dense_res = make_temporary_conversion<remove_complex<ValueType>>(
+        local_result.get());
+    exec->run(dense::make_compute_squared_norm2(this, dense_res.get(), tmp));
+}
+
+
+template <typename ValueType>
+void Dense<ValueType>::compute_squared_norm2_impl(LinOp* result) const
+{
+    GKO_ASSERT_EQUAL_DIMENSIONS(result, dim<2>(1, this->get_size()[1]));
+    auto exec = this->get_executor();
+    auto dense_res =
+        make_temporary_conversion<remove_complex<ValueType>>(result);
+    array<char> tmp{exec};
+    exec->run(dense::make_compute_squared_norm2(this, dense_res.get(), tmp));
+}
```

Comment:
this part can be marged into the `compute_squared_norm2(LinOp* result)`
it does not do the same thing as `compute_squared_norm2(LinOp* result, array<char>)`
I will suggest create the tmp in  `compute_squared_norm2(LinOp* result)` and call `compute_squared_norm2(result, tmp)`

Reply:
see the below comment together